README
A mod to Metacity to allow the user to "toss" windows, which bounce of
edges of screens. Currently only has been tested on single screen
setup, and doesn't work very well yet. For now, it assumes two 25px
bars, one on top of screen one on bottom.

This git repository is based on Metacity 2.28.0 (current Ubuntu 9.10
apt-get source), and will probably work on any fairly similar Metacity
versions.

Future plans include:
* Allowing windows to move between screens and/or workspaces
* Auto detection of taskbars
* Cleaning up coding to follow Metacity conventions
* Moving coding into other files for neatness
* Creating a GUI interface for all variables, plus an easy way to enable
   or disable the mod.
* Setting up an easy way to install mod

To install, follow the instructions from Metacity. Use at your own risk!

Mod by Alex Willisson <alex.willisson@gmail.com>


debian/changelog
metacity (1:2.28.0-0ubuntu1) karmic; urgency=low

  * New upstream release: (LP: #434933)
    - Translation updates

 -- Robert Ancell <robert.ancell@canonical.com>  Wed, 23 Sep 2009 09:46:49 +1000

metacity (1:2.27.1-0ubuntu1) karmic; urgency=low

  [ Robert Ancell ]
  * New upstream version (LP: #427195):
    - Unset _NET_SUPPORTING_WM_CHECK when shutting down
    - Add a switch to disable autofullscreen'ing maximized windows without
      decorations
    - Translation updates

  [ Martin Pitt ]
  * debian/control.in: Update Vcs-Bzr: for updated branch owner
    (~ubuntu-desktop now).

 -- Robert Ancell <robert.ancell@canonical.com>  Thu, 10 Sep 2009 20:15:44 +0200

metacity (1:2.27.0-0ubuntu4) karmic; urgency=low

  * debian/patches/015_set_restartifrunning_on_replace.patch:
    - Make sure that RestartStyleHint is set to RestartIfRunning
      when replaced (LP: #389686). 
    - Only display warning dialog for clients that don't implement session 
      saving during a "Local" or "Both" SaveYourself request when the user
      wants to save their session state, rather than at the end of every 
      session (LP: #35316). 
    - Patch taken from GNOME Bug #588119.

 -- Chris Coulson <chrisccoulson@ubuntu.com>  Wed, 09 Sep 2009 08:59:29 +0200

metacity (1:2.27.0-0ubuntu3) karmic; urgency=low

  * debian/patches/014_fix_wm_cm_selection_timestamp.patch:
    - get timestamp when starting and use it to get and release
      the _NET_WM_CM_SX selection (LP: #178953)

 -- Travis Watkins <amaranth@ubuntu.com>  Mon, 24 Aug 2009 05:11:47 -0500

metacity (1:2.27.0-0ubuntu2) karmic; urgency=low

  * debian/control.in:
  * debian/control:
    - Change build dependency from libcanberra-dev to libcanberra-gtk-dev to fix FTBFS (LP: #388713)

 -- Robert Ancell <robert.ancell@canonical.com>  Thu, 18 Jun 2009 09:03:13 +1000

metacity (1:2.27.0-0ubuntu1) karmic; urgency=low

  * New upstream version (LP: #388196):
    - Remove deprecated GTK symbols (Thomas A, Matt) (#572332)
    - fix window properties bug which showed up as ignoring struts (Ori) (#572573) (LP: #341487)
    - include "as <username>" in window title (Thomas T) (#549389)
    - Use zenity for dialogues (Thomas T) (#521914)
    - Use libcanberra for sound effects (Matthias) (#557921)
    - Don't define meta_spew_event unless verbose mode is on (Matt) (#571126)
    - Fix README to give correct switch names (Matt) (#571210)
    - Windows which couldn't appear in alt-tab don't get there by being urgent (Thomas T)
    - Skip the first window in alt-tab only if it's actually current (Matt) (#535887)
    - Optimise line coordinate storage (Thomas)
    - Queue resize on window undecorate (Neil)
    - Fix description of "show desktop" (Luca) (#569649)
    - Wrap g_error calls in braces (Matt)
  * debian/control.in:
  * debian/control:
    - Add build dependency on libcanberra-dev
  * debian/patches/014_fix_panel_behaviour.patch: Obsolete
  * debian/metacity.install: No longer have /usr/lib/metacity/metacity-dialog

 -- Robert Ancell <robert.ancell@canonical.com>  Wed, 17 Jun 2009 10:23:16 +0200

metacity (1:2.26.0-0ubuntu1) karmic; urgency=low

  * New upstream stable release; bug fixes and translation updates.
    - Properly resize windows after maximizing them; LP: #283447.

 -- Loic Minier <lool@dooz.org>  Mon, 11 May 2009 23:01:21 +0200

metacity (1:2.25.144-0ubuntu2) jaunty; urgency=low

  * debian/patches/014_fix_panel_behavior.patch
   - From upstream, fix windows behavior with various panels situations 
     like autohide mode for gnome-panel or "don't cover the bar" option of 
     avant-window-navigator. (LP: #341487)

 -- Julien Lavergne <julien.lavergne@gmail.com>  Mon, 30 Mar 2009 10:58:21 +0200

metacity (1:2.25.144-0ubuntu1) jaunty; urgency=low

  * New upstream version (LP: #324110):
    - Optimise the window property lookup.
    - Several memory leaks fixed.
    - Fix longstanding crasher about colormaps.
    - Alt+middle/right buttons can be switched.
  * 99_ltmain_as-needed.patch:
    - New version update.

 -- Chris Coulson <chrisccoulson@googlemail.com>  Sun, 01 Feb 2009 22:05:17 +0000

metacity (1:2.25.89-0ubuntu1) jaunty; urgency=low

    * New upstream version (LP: #312522)
      - The maximisation key is a toggle.

src/core/place.h
/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */

/* Metacity window placement */

/* 
 * Copyright (C) 2001 Havoc Pennington
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */

#ifndef META_PLACE_H
#define META_PLACE_H

#include "window-private.h"
#include "frame-private.h"

void meta_window_place (MetaWindow *window,
                        MetaFrameGeometry *fgeom,
                        int         x,
                        int         y,
                        int        *new_x,
                        int        *new_y);

#endif


src/include/window.h
/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */

/*
 * Copyright (C) 2008 Iain Holmes
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */

#ifndef META_WINDOW_H
#define META_WINDOW_H

#include <glib.h>
#include <X11/Xlib.h>

#include "boxes.h"
#include "types.h"

MetaFrame *meta_window_get_frame (MetaWindow *window);
gboolean meta_window_has_focus (MetaWindow *window);
gboolean meta_window_is_shaded (MetaWindow *window);
MetaRectangle *meta_window_get_rect (MetaWindow *window);
MetaScreen *meta_window_get_screen (MetaWindow *window);
MetaDisplay *meta_window_get_display (MetaWindow *window);
Window meta_window_get_xwindow (MetaWindow *window);

void alex_window_get_position (MetaWindow *window, int *x, int *y);

#endif


src/core/main.c
/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */

/* Metacity main() */

/* 
 * Copyright (C) 2001 Havoc Pennington
 * Copyright (C) 2006 Elijah Newren
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */

/**
 * \file 
 * Program startup.
 * Functions which parse the command-line arguments, create the display,
 * kick everything off and then close down Metacity when it's time to go.
 */

/**
 * \mainpage
 * Metacity - a boring window manager for the adult in you
 *
 * Many window managers are like Marshmallow Froot Loops; Metacity
 * is like Cheerios.
 *
 * The best way to get a handle on how the whole system fits together
 * is discussed in doc/code-overview.txt; if you're looking for functions
 * to investigate, read main(), meta_display_open(), and event_callback().
 */

#define _GNU_SOURCE
#define _SVID_SOURCE /* for putenv() and some signal-related functions */

#include <config.h>
#include "main.h"
#include "util.h"
#include "display-private.h"
#include "errors.h"
#include "ui.h"
#include "session.h"
#include "prefs.h"

#include <glib-object.h>
#include <glib/gprintf.h>

#include <stdlib.h>
#include <sys/types.h>
#include <wait.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <locale.h>
#include <time.h>

/**
 * The exit code we'll return to our parent process when we eventually die.
 */
static MetaExitCode meta_exit_code = META_EXIT_SUCCESS;

/**
 * Handle on the main loop, so that we have an easy way of shutting Metacity
 * down.
 */
static GMainLoop *meta_main_loop = NULL;

/**
 * If set, Metacity will spawn an identical copy of itself immediately
 * before quitting.
 */
static gboolean meta_restart_after_quit = FALSE;

static void prefs_changed_callback (MetaPreference pref,
                                    gpointer       data);

/**
 * Prints log messages. If Metacity was compiled with backtrace support,
 * also prints a backtrace (see meta_print_backtrace()).
 *
 * \param log_domain  the domain the error occurred in (we ignore this)
 * \param log_level   the log level so that we can filter out less
 *                    important messages
 * \param message     the message to log
 * \param user_data   arbitrary data (we ignore this)
 */
static void
log_handler (const gchar   *log_domain,
             GLogLevelFlags log_level,
             const gchar   *message,
             gpointer       user_data)
{
  meta_warning ("Log level %d: %s\n", log_level, message);
  meta_print_backtrace ();
}

/**
 * Prints the version notice. This is shown when Metacity is called
 * with the --version switch.
 */
static void
version (void)
{
  const int latest_year = 2009;
  char yearbuffer[256];
  GDate date;

  /* this is all so the string to translate stays constant.
   * see how much we love the translators.
   */
  g_date_set_dmy (&date, 1, G_DATE_JANUARY, latest_year);
  if (g_date_strftime (yearbuffer, sizeof (yearbuffer), "%Y", &date)==0)
    /* didn't work?  fall back to decimal representation */
    g_sprintf (yearbuffer, "%d", latest_year);

  g_print (_("metacity %s\n"
             "Copyright (C) 2001-%s Havoc Pennington, Red Hat, Inc., and others\n"
             "This is free software; see the source for copying conditions.\n"
             "There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"),
           VERSION, yearbuffer);
  exit (0);
}

/**
 * Prints a list of which configure script options were used to
 * build this copy of Metacity. This is actually always called
 * on startup, but it's all no-op unless we're in verbose mode
 * (see meta_set_verbose).
 */
static void
meta_print_compilation_info (void)
{
#ifdef HAVE_SHAPE
  meta_verbose ("Compiled with shape extension\n");
#else
  meta_verbose ("Compiled without shape extension\n");
#endif
#ifdef HAVE_XINERAMA
  meta_topic (META_DEBUG_XINERAMA, "Compiled with Xinerama extension\n");
#else
  meta_topic (META_DEBUG_XINERAMA, "Compiled without Xinerama extension\n");
#endif
#ifdef HAVE_XFREE_XINERAMA
  meta_topic (META_DEBUG_XINERAMA, " (using XFree86 Xinerama)\n");
#else
  meta_topic (META_DEBUG_XINERAMA, " (not using XFree86 Xinerama)\n");
#endif
#ifdef HAVE_SOLARIS_XINERAMA
  meta_topic (META_DEBUG_XINERAMA, " (using Solaris Xinerama)\n");
#else
  meta_topic (META_DEBUG_XINERAMA, " (not using Solaris Xinerama)\n");
#endif
#ifdef HAVE_XSYNC
  meta_verbose ("Compiled with sync extension\n");
#else
  meta_verbose ("Compiled without sync extension\n");
#endif
#ifdef HAVE_RANDR
  meta_verbose ("Compiled with randr extension\n");
#else
  meta_verbose ("Compiled without randr extension\n");
#endif
#ifdef HAVE_STARTUP_NOTIFICATION
  meta_verbose ("Compiled with startup notification\n");
#else
  meta_verbose ("Compiled without startup notification\n");
#endif
#ifdef HAVE_COMPOSITE_EXTENSIONS
  meta_verbose ("Compiled with composite extensions\n");
#else
  meta_verbose ("Compiled without composite extensions\n");
#endif
}

/**
 * Prints the version number, the current timestamp (not the
 * build date), the locale, the character encoding, and a list
 * of configure script options that were used to build this
 * copy of Metacity. This is actually always called
 * on startup, but it's all no-op unless we're in verbose mode
 * (see meta_set_verbose).
 */
static void
meta_print_self_identity (void)
{
  char buf[256];
  GDate d;
  const char *charset;

  /* Version and current date. */
  g_date_clear (&d, 1);
  g_date_set_time_t (&d, time (NULL));
  g_date_strftime (buf, sizeof (buf), "%x", &d);
  meta_verbose ("Metacity version %s running on %s\n",
    VERSION, buf);
  
  /* Locale and encoding. */
  g_get_charset (&charset);
  meta_verbose ("Running in locale \"%s\" with encoding \"%s\"\n",
    setlocale (LC_ALL, NULL), charset);

  /* Compilation settings. */
  meta_print_compilation_info ();
}

/**
 * The set of possible options that can be set on Metacity's
 * command line. This type exists so that meta_parse_options() can
 * write to an instance of it.
 */
typedef struct
{
  gchar *save_file;
  gchar *display_name;
  gchar *client_id;
  gboolean replace_wm;
  gboolean disable_sm;
  gboolean print_version;
  gboolean sync;
  gboolean composite;
  gboolean no_composite;
  gboolean no_force_fullscreen;
} MetaArguments;

#ifdef HAVE_COMPOSITE_EXTENSIONS
#define COMPOSITE_OPTS_FLAGS 0
#else /* HAVE_COMPOSITE_EXTENSIONS */
/* No compositor, so don't show the arguments in --help */
#define COMPOSITE_OPTS_FLAGS G_OPTION_FLAG_HIDDEN
#endif /* HAVE_COMPOSITE_EXTENSIONS */

/**
 * Parses argc and argv and returns the
 * arguments that Metacity understands in meta_args.
 *
 * The strange call signature has to be written like it is so
 * that g_option_context_parse() gets a chance to modify argc and
 * argv.
 *
 * \param argc  Pointer to the number of arguments Metacity was given
 * \param argv  Pointer to the array of arguments Metacity was given
 * \param meta_args  The result of parsing the arguments.
 **/
static void
meta_parse_options (int *argc, char ***argv,
                    MetaArguments *meta_args)
{
  MetaArguments my_args = {NULL, NULL, NULL,
                           FALSE, FALSE, FALSE, FALSE, FALSE};
  GOptionEntry options[] = {
    {
      "sm-disable", 0, 0, G_OPTION_ARG_NONE,
      &my_args.disable_sm,
      N_("Disable connection to session manager"),
      NULL
    },
    {
      "replace", 0, 0, G_OPTION_ARG_NONE,
      &my_args.replace_wm,
      N_("Replace the running window manager with Metacity"),
      NULL
    },
    {
      "sm-client-id", 0, 0, G_OPTION_ARG_STRING,
      &my_args.client_id,
      N_("Specify session management ID"),
      "ID"
    },
    {
      "display", 'd', 0, G_OPTION_ARG_STRING,
      &my_args.display_name, N_("X Display to use"),
      "DISPLAY"
    },
    {
      "sm-save-file", 0, 0, G_OPTION_ARG_FILENAME,
      &my_args.save_file,
      N_("Initialize session from savefile"),
      "FILE"
    },
    {
      "version", 0, 0, G_OPTION_ARG_NONE,
      &my_args.print_version,
      N_("Print version"),
      NULL
    },
    {
      "sync", 0, 0, G_OPTION_ARG_NONE,
      &my_args.sync,
      N_("Make X calls synchronous"),
      NULL
    },
    {
      "composite", 'c', COMPOSITE_OPTS_FLAGS, G_OPTION_ARG_NONE,
      &my_args.composite,
      N_("Turn compositing on"),
      NULL
    },
    {
      "no-composite", 0, COMPOSITE_OPTS_FLAGS, G_OPTION_ARG_NONE,
      &my_args.no_composite,
      N_("Turn compositing off"),
      NULL
    },
    {
      "no-force-fullscreen", 0, COMPOSITE_OPTS_FLAGS, G_OPTION_ARG_NONE,
      &my_args.no_force_fullscreen,
      N_("Don't make fullscreen windows that are maximized and have no decorations"),
      NULL
    },
    {NULL}
  };
  GOptionContext *ctx;
  GError *error = NULL;

  ctx = g_option_context_new (NULL);
  g_option_context_add_main_entries (ctx, options, "metacity");
  if (!g_option_context_parse (ctx, argc, argv, &error))
    {
      g_print ("metacity: %s\n", error->message);
      exit(1);
    }
  g_option_context_free (ctx);
  /* Return the parsed options through the meta_args param. */
  *meta_args = my_args;
}

/**
 * Selects which display Metacity should use. It first tries to use
 * display_name as the display. If display_name is NULL then
 * try to use the environment variable METACITY_DISPLAY. If that
 * also is NULL, use the default - :0.0
 */
static void
meta_select_display (gchar *display_name)
{
  gchar *envVar = "";
  if (display_name)
    envVar = g_strconcat ("DISPLAY=", display_name, NULL);
  else if (g_getenv ("METACITY_DISPLAY"))
    envVar = g_strconcat ("DISPLAY=",
      g_getenv ("METACITY_DISPLAY"), NULL);
  /* DO NOT FREE envVar, putenv() sucks */
  putenv (envVar);
}

static void
meta_finalize (void)
{
  MetaDisplay *display = meta_get_display();

  meta_session_shutdown ();

  if (display)
    meta_display_close (display,
                        CurrentTime); /* I doubt correct timestamps matter here */
}

static void
sigterm_handler (int signum)
{
  meta_finalize ();

  exit (meta_exit_code);
}

static guint sigchld_signal_id = 0;

static void
sigchld_handler (int signum, siginfo_t *info, void *context)
{
  int stat;

  if (info->si_code == CLD_EXITED)
    {
      g_signal_emit (sigchld_nexus, sigchld_signal_id, 0,
                     info->si_status,
                     GINT_TO_POINTER (info->si_pid));
    }

  g_signal_handlers_disconnect_matched (sigchld_nexus,
                                        G_SIGNAL_MATCH_DATA,
                                        sigchld_signal_id,
                                        0, NULL, NULL,
                                        GINT_TO_POINTER (info->si_pid));

  waitpid (info->si_pid, &stat, WNOHANG);
}
    
/**
 * This is where the story begins. It parses commandline options and
 * environment variables, sets up the screen, hands control off to
 * GTK, and cleans up afterwards.
 *
 * \param argc Number of arguments (as usual)
 * \param argv Array of arguments (as usual)
 *
 * \bug It's a bit long. It would be good to split it out into separate
 * functions.
 */

#include <time.h>
#include <sys/time.h>
#include <math.h>
#include "frame-private.h"

struct vect {
  double x, y;
};

static double
get_secs (void)
{
  struct timeval tv;
  gettimeofday (&tv, NULL);
  return (tv.tv_sec + tv.tv_usec/1e6);
}

static void
alex_window (gpointer spec, gpointer user_data)
{
  MetaWindow *window = spec;
  MetaRectangle rect;
  MetaFrameGeometry geomp;
  double now, dt, dx, dy, inst_vel, time_const, factor;
  int left, top, right, bottom, newx, newy, screenx, screeny, bounce_flag;
  static double velx, vely;

  now = get_secs ();
  dt = now - window->lasttime;

  window->lasttime = now;

  meta_window_get_client_root_coords (window, &rect);

  alex_window_get_position (window, &left, &top);

  if (window->frame)
    meta_frame_calc_geometry (window->frame, &geomp);
  else
    return;

  right = left + geomp.left_width + rect.width + geomp.right_width;
  bottom = top + geomp.top_height + rect.height + geomp.bottom_height;

  switch (window->phys_state) {
  case 1:
    time_const = dt;
    
    dx = rect.x - window->lastx;
    inst_vel = (dx / dt);
    factor = dt / time_const;
    if (factor < 0)
      factor = 0;
    if (factor > 1)
      factor = 1;
    velx = velx * (1 - factor) + inst_vel * factor;

    dy = rect.y - window->lasty;
    inst_vel = (dy / dt);
    factor = dt / time_const;
    if (factor < 0)
      factor = 0;
    if (factor > 1)
      factor = 1;
    vely = vely * (1 - factor) + inst_vel * factor;

    window->theta = atan2 (vely, velx);
    window->speed = hypot (vely, velx);
    break;
  case 2:
    if (now - window->lasttimemouse > .1) {
      window->speed = 0;
      window->velx = 0;
      window->vely = 0;
      window->phys_state = 0;
      break;
    }

    window->phys_state = 3;
  case 3:
    bounce_flag = 0;

    velx = window->speed * cos (window->theta);
    vely = window->speed * sin (window->theta);

    newx = rect.x + velx * dt;
    newy = rect.y + vely * dt;

    if (left + velx * dt <= 0) {
      meta_window_move (window, TRUE, 0, newy);

      window->theta = atan2 (vely, -velx);
      window->speed -= 200;

      bounce_flag = 1;
    }

    if (top + vely * dt <= 25) {
      meta_window_move (window, TRUE, newx, 25 + geomp.top_height);

      window->theta = atan2 (-vely, velx);
      window->speed -= 200;

      bounce_flag = 1;
    }

   screenx = window->screen->xscreen->width;
   screeny = window->screen->xscreen->height;

    if (right + velx * dt >= screenx) {
      meta_window_move (window, TRUE, screenx - geomp.right_width - rect.width, newy);

      window->theta = atan2 (vely, -velx);
      window->speed -= 200;

      bounce_flag = 1;
    }

    if (bottom + vely * dt >= screeny - 25) {
      meta_window_move (window, TRUE, newx, screeny - 25 - geomp.bottom_height - rect.height);

      window->theta = atan2 (-vely, velx);
      window->speed -= 200;

      bounce_flag = 1;
    }

    if (bounce_flag == 0) {
      meta_window_move (window, TRUE, newx, newy);
    }

    if (window->speed < 20) {
      window->speed = 0;
      window->phys_state = 0;
    }
    
    window->speed -= 1000 * dt;
    break;
  }
}

static gboolean
alex_testing (gpointer data)
{
  MetaDisplay *alexdisplay;
  GSList *display_windows;

  alexdisplay = meta_get_display ();

  display_windows = meta_display_list_windows (alexdisplay);

  g_slist_foreach (display_windows, alex_window, NULL);

  return (1);
}

int
main (int argc, char **argv)
{
  struct sigaction act;
  sigset_t empty_mask;
  MetaArguments meta_args;
  const gchar *log_domains[] = {
    NULL, G_LOG_DOMAIN, "Gtk", "Gdk", "GLib",
    "Pango", "GLib-GObject", "GThread"
  };
  guint i;

  if (!g_thread_supported ())
    g_thread_init (NULL);
  
  if (setlocale (LC_ALL, "") == NULL)
    meta_warning ("Locale not understood by C library, internationalization will not work\n");

  g_type_init ();
  
  sigemptyset (&empty_mask);
  act.sa_handler = SIG_IGN;
  act.sa_mask    = empty_mask;
  act.sa_flags   = 0;
  if (sigaction (SIGPIPE,  &act, NULL) < 0)
    g_printerr ("Failed to register SIGPIPE handler: %s\n",
                g_strerror (errno));
#ifdef SIGXFSZ
  if (sigaction (SIGXFSZ,  &act, NULL) < 0)
    g_printerr ("Failed to register SIGXFSZ handler: %s\n",
                g_strerror (errno));
#endif

  act.sa_handler = &sigterm_handler;
  if (sigaction (SIGTERM, &act, NULL) < 0)
    g_printerr ("Failed to register SIGTERM handler: %s\n",
		g_strerror (errno));

  sigchld_nexus = g_object_new (META_TYPE_NEXUS, NULL);

  sigchld_signal_id =
    g_signal_new ("sigchld", META_TYPE_NEXUS,
                  G_SIGNAL_RUN_LAST,
                  0, NULL, NULL,
                  g_cclosure_marshal_VOID__UINT_POINTER,
                  G_TYPE_NONE,
                  2,
                  G_TYPE_UINT, G_TYPE_POINTER);
  
  act.sa_flags = SA_NOCLDSTOP | SA_SIGINFO;
  act.sa_handler = SIG_DFL;
  act.sa_sigaction = &sigchld_handler;
  if (sigaction (SIGCHLD, &act, NULL) < 0)
    g_printerr ("Failed to register SIGCHLD handler: %s\n",
		g_strerror (errno));

  if (g_getenv ("METACITY_VERBOSE"))
    meta_set_verbose (TRUE);
  if (g_getenv ("METACITY_DEBUG"))
    meta_set_debugging (TRUE);

  if (g_get_home_dir ())
    if (chdir (g_get_home_dir ()) < 0)
      meta_warning ("Could not change to home directory %s.\n",
                    g_get_home_dir ());

  meta_print_self_identity ();
  
  bindtextdomain (GETTEXT_PACKAGE, METACITY_LOCALEDIR);
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  textdomain (GETTEXT_PACKAGE);

  /* Parse command line arguments.*/
  meta_parse_options (&argc, &argv, &meta_args);

  meta_set_syncing (meta_args.sync || (g_getenv ("METACITY_SYNC") != NULL));

  if (meta_args.print_version)
    version ();

  meta_select_display (meta_args.display_name);
  
  if (meta_args.replace_wm)
    meta_set_replace_current_wm (TRUE);

  if (meta_args.save_file && meta_args.client_id)
    meta_fatal ("Can't specify both SM save file and SM client id\n");
  
  meta_main_loop = g_main_loop_new (NULL, FALSE);
  
  g_timeout_add (30, alex_testing, NULL);

  meta_ui_init (&argc, &argv);  

  /* must be after UI init so we can override GDK handlers */
  meta_errors_init ();

  /* Load prefs */
  meta_prefs_init ();
  meta_prefs_add_listener (prefs_changed_callback, NULL);


#if 1

  for (i=0; i<G_N_ELEMENTS(log_domains); i++)
    g_log_set_handler (log_domains[i],
                       G_LOG_LEVEL_MASK | G_LOG_FLAG_FATAL | G_LOG_FLAG_RECURSION,
                       log_handler, NULL);

#endif

  if (g_getenv ("METACITY_G_FATAL_WARNINGS") != NULL)
    g_log_set_always_fatal (G_LOG_LEVEL_MASK);
  
  meta_ui_set_current_theme (meta_prefs_get_theme (), FALSE);

  /* Try to find some theme that'll work if the theme preference
   * doesn't exist.  First try Simple (the default theme) then just
   * try anything in the themes directory.
   */
  if (!meta_ui_have_a_theme ())
    meta_ui_set_current_theme ("Simple", FALSE);
  
  if (!meta_ui_have_a_theme ())
    {
      const char *dir_entry = NULL;
      GError *err = NULL;
      GDir   *themes_dir = NULL;
      
      if (!(themes_dir = g_dir_open (METACITY_DATADIR"/themes", 0, &err)))
        {
          meta_fatal (_("Failed to scan themes directory: %s\n"), err->message);
          g_error_free (err);
        } 
      else 
        {
          while (((dir_entry = g_dir_read_name (themes_dir)) != NULL) && 
                 (!meta_ui_have_a_theme ()))
            {
              meta_ui_set_current_theme (dir_entry, FALSE);
            }
          
          g_dir_close (themes_dir);
        }
    }
  
  if (!meta_ui_have_a_theme ())
    meta_fatal (_("Could not find a theme! Be sure %s exists and contains the usual themes.\n"),
                METACITY_DATADIR"/themes");
  
  /* Connect to SM as late as possible - but before managing display,
   * or we might try to manage a window before we have the session
   * info
   */
  if (!meta_args.disable_sm)
    {
      if (meta_args.client_id == NULL)
        {
          const gchar *desktop_autostart_id;
  
          desktop_autostart_id = g_getenv ("DESKTOP_AUTOSTART_ID");
 
          if (desktop_autostart_id != NULL)
            meta_args.client_id = g_strdup (desktop_autostart_id);
        }

      /* Unset DESKTOP_AUTOSTART_ID in order to avoid child processes to
       * use the same client id. */
      g_unsetenv ("DESKTOP_AUTOSTART_ID");

      meta_session_init (meta_args.client_id, meta_args.save_file);
    }
  /* Free memory possibly allocated by the argument parsing which are
   * no longer needed.
   */
  g_free (meta_args.save_file);
  g_free (meta_args.display_name);
  g_free (meta_args.client_id);

  if (meta_args.composite || meta_args.no_composite)
    meta_prefs_set_compositing_manager (meta_args.composite);

  if (meta_args.no_force_fullscreen)
    meta_prefs_set_force_fullscreen (FALSE);

  if (!meta_display_open ())
    meta_exit (META_EXIT_ERROR);
  
  g_main_loop_run (meta_main_loop);

  meta_finalize ();

  if (meta_restart_after_quit)
    {
      GError *err;

      err = NULL;
      if (!g_spawn_async (NULL,
                          argv,
                          NULL,
                          G_SPAWN_SEARCH_PATH,
                          NULL,
                          NULL,
                          NULL,
                          &err))
        {
          meta_fatal (_("Failed to restart: %s\n"),
                      err->message);
          g_error_free (err); /* not reached anyhow */
          meta_exit_code = META_EXIT_ERROR;
        }
    }
  
  return meta_exit_code;
}

/**
 * Stops Metacity. This tells the event loop to stop processing; it is rather
 * dangerous to use this rather than meta_restart() because this will leave
 * the user with no window manager. We generally do this only if, for example,
 * the session manager asks us to; we assume the session manager knows what
 * it's talking about.
 *
 * \param code The success or failure code to return to the calling process.
 */
void
meta_quit (MetaExitCode code)
{
  meta_exit_code = code;

  if (g_main_loop_is_running (meta_main_loop))
    g_main_loop_quit (meta_main_loop);
}

/**
 * Restarts Metacity. In practice, this tells the event loop to stop
 * processing, having first set the meta_restart_after_quit flag which
 * tells Metacity to spawn an identical copy of itself before quitting.
 * This happens on receipt of a _METACITY_RESTART_MESSAGE client event.
 */
void
meta_restart (void)
{
  meta_restart_after_quit = TRUE;
  meta_quit (META_EXIT_SUCCESS);
}

/**
 * Called on pref changes. (One of several functions of its kind and purpose.)
 *
 * \bug Why are these particular prefs handled in main.c and not others?
 * Should they be?
 *
 * \param pref  Which preference has changed
 * \param data  Arbitrary data (which we ignore)
 */
static void
prefs_changed_callback (MetaPreference pref,
                        gpointer       data)
{
  switch (pref)
    {
    case META_PREF_THEME:
      meta_ui_set_current_theme (meta_prefs_get_theme (), FALSE);
      meta_display_retheme_all ();
      break;

    case META_PREF_CURSOR_THEME:
    case META_PREF_CURSOR_SIZE:
      meta_display_set_cursor_theme (meta_prefs_get_cursor_theme (),
				     meta_prefs_get_cursor_size ());
      break;
    default:
      /* handled elsewhere or otherwise */
      break;
    }
}
